src/synapse-plugins/calculator-plugin/calculator-plugin-backend.vala

```diff
@@ -30,81 +30,67 @@ namespace Synapse {
             return instance;
         }
 
-        private Regex? express_regex = null;
-        private Regex? base_regex = null;
         private static CalculatorPluginBackend instance = null;
-
-        construct {
-            try {
-                /* The express_regex describes a string which *resembles* a mathematical expression in one of two forms:
-                <alphanum><operator><alphanum> e.g. 2 + 2
-                <opening parenthesis><number expression><closing parenthesis) e.g. sqrt (0.5)
-                */
-                express_regex = new Regex (
-                    """^.*(\w+[\/\+\-\*\^\%\!\&\|]{1,2}\.?\w+|\(-?\d+.*\))+.*$""",
-                    RegexCompileFlags.OPTIMIZE
-                );
-                /* The base_regex describes a string which starts with a bc number base expression */
-                base_regex = new Regex (
-                    """^.base=\d+;.*$""",
-                    RegexCompileFlags.OPTIMIZE
-                );
-            } catch (Error e) {
-                critical ("Error creating regexp: %s", e.message);
-                assert_not_reached ();
-            }
-        }
+        private const string VALID_NUM = "0123456789ABCDEF";
+        private const string VALID_OP = ".()%^&|!*/-+";
+        private string use_num;
 
         public async double get_solution (string query_string, Cancellable cancellable) throws Error {
             string? solution = null;
-            string input = query_string.replace (" ", "").replace (",", ".").replace ("x", "*");
-            // Mark characters not allowed in simple bc expressions
-            bool matched = true;
-            if (base_regex.match (input)) {
-                // If a number base is set, the expression may include hexadecimals
-                // or be doing a conversion, in which there is no expression
-                // so omit regex test and instead limit to certain characters for simple expressions
-                input.canon ("1234567890ABCDEF();%^&|!*/-+iobase=.", '@');
-
-            } else {
-                // Disallow capitals and test for possible mathematical expression
-                input = input.down ();
-                matched = express_regex.match (input);
+            // Assume base 10 unless indicated otherwise
+            use_num = VALID_NUM.slice (0, 10);
+            var input = query_string.replace (" ", "")
+                        .replace (",", ".")
+                        .replace ("exp(", "e(") // Must do before replacing "x"
+                        .replace ("x", "*")
+                        .replace ("ln(", "l(")
+                        .replace ("sin(", "s(")
+                        .replace ("cos(", "c(")
+                        .replace ("pi", "(a(1)*4)")
+                        .replace ("π", "(a(1)*4)");
+            string[] expressions = input.split (";");
```

Comment:
Should we add more unit tests with ;'s usage? There currently is only one. 
It seems only relevant for usage with `scale`, `ibase` and `obase`. Might be interesting to clarify that in a comment?

This is a bit subjective, but should we even support this advanced bc functionality in the applications menu?

Reply:
As you say, subjective.  It could be argued that an applications menu is no place for a calculator (or unit convertor) anyway but if we are to support it at all it seemed reasonable we should support as much as can be written in one line.  I suspect this functionality is little used in practice as it is not very discoverable.

I'll work on more unit tests.  You are right that if we support it then it should be robust.

Reply:
I've added some more tests (and fixed handling of "obase=").  Let me know if any additional or better tests needed.

Reply:
Looks great :+1: